### PR TITLE
Return INVALID instead of aborting in type kinds

### PR DESCRIPTION
**Pull request by eboasson**
If #2105 gets merged, which I think it should, this PR's changes still make sense as a matter of principle but there is then no known way to trigger the abort.

### CODE REVIEW

**Review by dpotman — COMMENTED** [review]
Looks good, only a detail in the error handling in `add_type_info_hash_deps_r` (see below)

File: src/core/ddsi/src/ddsi_typelib.c [hunk 1]
```diff
@@ -1038,6 +1038,10 @@ static dds_return_t add_type_info_hash_deps_r (struct ddsi_domaingv *gv, const d
 
       case DDSI_TYPEID_KIND_FULLY_DESCRIPTIVE:
         break;
+
+      case DDSI_TYPEID_KIND_INVALID:
+        ret = DDS_RETCODE_BAD_PARAMETER;
```
[the remark this answers is not in the document]
Comment (eboasson) [reply src/core/ddsi/src/ddsi_typelib.c:1043]: Good point, the code is indeed somewhat fishy.

If I'm not mistaken `ddsi_typeid_fini(x) ; ddsi_typeid_fini(x)` is allowed because you end up in `dds_stream_free_sample`, which nulls the pointers it freed and therefore calling it a second time will be a no-op. That suggests the `ddsi_typeid_fini (&tmpl.src_type_id)` is harmless for the minimal/complete cases.

Secondly, on a cursory glance, all callers of `add_type_info_hash_deps_r` free the `type_info` on error, except for recursive calls where it simply passes it unchanged, so that seems okay, too.

In short I wouldn't be surprised if it ended up doing the right thing, but if so, it does it in a manner so roundabout that I don't feel comfortable with it. In short, I'll go double-check the code paths and see if I can eliminate the weirdness without harming (apparent) readability 🙂

**Review by dpotman — APPROVED** [review]
LGTM
